tiatoolbox/tools/patchextraction.py

```diff
@@ -146,21 +177,163 @@ def _generate_location_df(self):
 
         self.locations_df = misc.read_locations(input_table=np.array(data))
 
         return self
 
+    @staticmethod
+    def filter_coordinates(
+        mask_reader, coordinates_list, func=None, resolution=None, units=None
+    ):
+        """
+        Indicates which coordinate is valid for mask-based patch extraction.
+        Locations are being validated by a custom or build-in `func`.
+
+        Args:
+            mask_reader (:class:`.VirtualReader`): a virtual pyramidal reader of the
+                mask related to the WSI from which we want to extract the patches.
+            coordinates_list (ndarray and np.int32): Coordinates to be checked
+                via the `func`. They must be in the same resolution as requested
+                `resolution` and `units`. The shape of `coordinates_list` is (N, K)
+                where N is the number of coordinate sets and K is either 2 for centroids
+                or 4 for bounding boxes. When using the default `func=None`, K should be
+                4, as we expect the `coordinates_list` to be refer to bounding boxes in
+                `[start_x, start_y, end_x, end_y]` format.
+            func: The coordinate validator function. A function that takes `reader` and
+                `coordinate` as arguments and return True or False as indication of
+                coordinate validity.
+
+        Returns:
+            ndarray: list of flags to indicate which coordinate is valid.
+        """
+
+        def default_sel_func(reader: wsireader.VirtualWSIReader, coord: np.ndarray):
+            """Accept coord as long as its box contains bits of mask."""
+            roi = reader.read_bounds(
+                coord,
+                resolution=reader.info.mpp if resolution is None else resolution,
+                units="mpp" if units is None else units,
+                interpolation="nearest",
+            )
+            return np.sum(roi > 0) > 0
+
+        if not isinstance(mask_reader, wsireader.VirtualWSIReader):
+            raise ValueError("`mask_reader` should be wsireader.VirtualWSIReader.")
+        if not isinstance(coordinates_list, np.ndarray) or not np.issubdtype(
+            coordinates_list.dtype, np.integer
+        ):
+            raise ValueError("`coordinates_list` should be ndarray of integer type.")
+        if func is None and coordinates_list.shape[-1] != 4:
+            raise ValueError(
+                "Default `func` does not support "
+                "`coordinates_list` of shape {}.".format(coordinates_list.shape)
+            )
+        func = default_sel_func if func is None else func
+        flag_list = [func(mask_reader, coord) for coord in coordinates_list]
+        return np.array(flag_list)
+
+    @staticmethod
+    def get_coordinates(
+        image_shape=None,
+        patch_shape=None,
+        stride_shape=None,
+        within_bound=True,
+    ):
+        """Calculate patch tiling coordinates.
```

Comment:
how is it different from _locations_df?

Reply:
`get_coordinates` is static function that is scale/resolution agnostic and can be used anywhere on demand. While `_generate_location_df` is class specific function that uses the former and other function and tricks to calculate resolution and region specific locations for patch extraction. Having these two functions separated allow us to expand this part more easily in the future.